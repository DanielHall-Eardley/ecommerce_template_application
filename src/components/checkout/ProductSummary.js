import React from 'react'
import styles from './ProductSummary.module.css'

export default ({removeProduct, order, user}) => {
  const products = order.products.map((product, index) => {
    return (
      <li key={product._id + index} className={styles.productItem}>
        <span>Product Id: {product._id}</span>
        <span>Name: {product.name}</span>
        <span className={styles.price}>${product.price}</span>
        { !order.addressConfirmed ?
          <button 
            className={styles.remove} 
            onClick={() => removeProduct(product._id, user.token, user.userId, order._id)}>
            Remove Product
          </button>
        : null }
      </li>
    )
  })

  return <ul className={styles.products} aria-label='products'>{products}</ul>
}