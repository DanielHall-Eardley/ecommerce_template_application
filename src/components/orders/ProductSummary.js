import React from 'react'
import styles from './ProductSummary.module.css'


/*This component renders a shipment list for a business 
admin user and a product list for a customer user. Only the
shipment list contains the links to print the postage labels*/
export default ({order, userType}) => {
  const renderProductOrShipment = order => {
    if (userType === 'admin' && !order.fulfilled) {
      return order.shipments.map(shipment => {
        return (
          <li key={shipment.shipmentId} className={styles.productItem} aria-label='shipment'>
            <span>Product Id: {shipment.productId}</span>
            <span>Name: {shipment.productName}</span>
            { shipment.postageLabel ?
              <a href={shipment.postageLabel} aria-live='polite'>Print Label</a>
            : null }  
          </li>
        )
      })
    } else {
      return order.products.map(product => {
        return (
          <li key={product.name} className={styles.productItem} aria-label='product'>
            <span>Product Id: {product._id}</span>
            <span>Name: {product.name}</span>
            <span>${product.price}</span>
          </li>
        )
      })
    }
  }

  return <ul className={styles.products} aria-label='product'>{renderProductOrShipment(order)}</ul>
}

