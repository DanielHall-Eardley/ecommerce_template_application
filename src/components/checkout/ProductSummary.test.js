import React from 'react'

import {shallow} from 'enzyme'
import {ProductSummary} from './ProductSummary.js'

let props;
let wrapper

const apiResponse = [
  {
    _id: 2,
    name: 'Second product',
    price: 350,
  }
]

beforeEach(() => {
  props = {
    clearError: jest.fn(),
    displayError: jest.fn(),
    authApi: jest.fn().mockResolvedValue(apiResponse),
    storeOrder: jest.fn(),
    order: {
      _id: 1,
      products: [
        {
          _id: 1,
          name: 'First product',
          price: 450,
        },
        {
          _id: 2,
          name: 'Second product',
          price: 350,
        },
      ],
      addressConfirmed: false
    },
    user: {
      userId: 1,
      token: 'Unique token'
    }
  }

  wrapper = shallow(<ProductSummary {...props} />)
})

test('test snapshot matches and is up to date', () => {
  expect(wrapper).toMatchSnapshot()
})

test('products cannot removed after address confirmation', () => {
  props.order.addressConfirmed = true
  wrapper = shallow(<ProductSummary {...props} />)
  expect(wrapper.exists('.remove')).toEqual(false)
})

test('products can be removed before address confirmation', () => {
  expect(wrapper.exists('.remove')).toEqual(true)
})

test('test the remove product functionality with a successful api response', async () => {
  const button = wrapper.find('ul').childAt(0).find('button')
  button.simulate('click')

  const body = {
    productId: 1,
    userId: props.user.userId,
    orderId: props.order._id
  }

  expect(props.clearError).toHaveBeenCalledTimes(1)
  expect(props.authApi).toBeCalledWith(
    '/checkout/remove-product',
    props.user.token,
    body,
    "PUT",
    props.displayError
  )
  await expect(props.authApi()).resolves.toBe(apiResponse)
  expect(props.storeOrder).toBeCalledWith(apiResponse)
})

test('test the remove product functionality with a failed api response', async () => {
  props.authApi = jest.fn().mockResolvedValue(false)
  wrapper = shallow(<ProductSummary {...props} />)
  const button = wrapper.find('ul').childAt(0).find('button')
  button.simulate('click')
 
  const body = {
    productId: 1,
    userId: props.user.userId,
    orderId: props.order._id
  }

  expect(props.clearError).toHaveBeenCalledTimes(1)
  expect(props.authApi).toBeCalledWith(
    '/checkout/remove-product',
    props.user.token,
    body,
    "PUT",
    props.displayError
  )
  await expect(props.authApi()).resolves.toBe(false)
})