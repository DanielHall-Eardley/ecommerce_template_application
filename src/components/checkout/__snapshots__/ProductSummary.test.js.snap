// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test snapshot matches and is up to date 1`] = `ShallowWrapper {}`;
